File: backend/internal/organization/mandate_test.go
package organization

import (
	"errors"
	"testing"
	"time"

	"github.com/google/uuid"
)

func ptr[T any](v T) *T { return &v }

func TestMandateStatus(t *testing.T) {
	now := time.Date(2026, 8, 11, 12, 0, 0, 0, time.UTC)
	hour := time.Hour

	tests := []struct {
		name string
		m    Mandate
		want string
	}{
		{
			name: "open-ended and in force is active",
			m:    Mandate{ValidFrom: now.Add(-hour)},
			want: MandateStatusActive,
		},
		{
			name: "window not open yet is pending",
			m:    Mandate{ValidFrom: now.Add(hour)},
			want: MandateStatusPending,
		},
		{
			name: "window closed is expired",
			m:    Mandate{ValidFrom: now.Add(-2 * hour), ValidUntil: ptr(now.Add(-hour))},
			want: MandateStatusExpired,
		},
		{
			name: "valid_until exactly now is expired",
			m:    Mandate{ValidFrom: now.Add(-hour), ValidUntil: ptr(now)},
			want: MandateStatusExpired,
		},
		{
			name: "revoked wins over expired",
			m: Mandate{
				ValidFrom:  now.Add(-2 * hour),
				ValidUntil: ptr(now.Add(-hour)),
				RevokedAt:  ptr(now.Add(-90 * time.Minute)),
			},
			want: MandateStatusRevoked,
		},
		{
			name: "revoked wins over active",
			m:    Mandate{ValidFrom: now.Add(-hour), RevokedAt: ptr(now.Add(-time.Minute))},
			want: MandateStatusRevoked,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := mandateStatus(tt.m, now); got != tt.want {
				t.Errorf("mandateStatus = %q, want %q", got, tt.want)
			}
		})
	}
}

func TestClampToParent(t *testing.T) {
	now := time.Date(2026, 8, 11, 12, 0, 0, 0, time.UTC)
	day := 24 * time.Hour
	dept := uuid.New()
	otherDept := uuid.New()

	activeFull := Mandate{
		Type:      MandateFull,
		Scope:     MandateScopeOrganization,
		ValidFrom: now.Add(-day),
	}

	t.Run("administrative under full keeps its own window", func(t *testing.T) {
		req := MandateGrant{
			Type: MandateAdministrative, Scope: MandateScopeOrganization,
			ValidFrom: now, ValidUntil: ptr(now.Add(day)),
		}
		got, err := clampToParent(req, activeFull, now)
		if err != nil {
			t.Fatalf("clampToParent: %v", err)
		}
		if !got.ValidFrom.Equal(now) || !got.ValidUntil.Equal(now.Add(day)) {
			t.Errorf("window = %v..%v, want unchanged", got.ValidFrom, got.ValidUntil)
		}
	})

	t.Run("full under administrative is over-delegation", func(t *testing.T) {
		parent := Mandate{Type: MandateAdministrative, Scope: MandateScopeOrganization, ValidFrom: now.Add(-day)}
		req := MandateGrant{Type: MandateFull, Scope: MandateScopeOrganization, ValidFrom: now}
		if _, err := clampToParent(req, parent, now); !errors.Is(err, ErrOverDelegation) {
			t.Errorf("err = %v, want ErrOverDelegation", err)
		}
	})

	t.Run("unknown type cannot be delegated", func(t *testing.T) {
		req := MandateGrant{Type: "owner", Scope: MandateScopeOrganization, ValidFrom: now}
		if _, err := clampToParent(req, activeFull, now); !errors.Is(err, ErrOverDelegation) {
			t.Errorf("err = %v, want ErrOverDelegation", err)
		}
	})

	t.Run("window is clamped to the parent's end", func(t *testing.T) {
		parent := activeFull
		parent.ValidUntil = ptr(now.Add(day))
		req := MandateGrant{
			Type: MandateAdministrative, Scope: MandateScopeOrganization,
			ValidFrom: now, ValidUntil: ptr(now.Add(10 * day)),
		}
		got, err := clampToParent(req, parent, now)
		if err != nil {
			t.Fatalf("clampToParent: %v", err)
		}
		if !got.ValidUntil.Equal(now.Add(day)) {
			t.Errorf("validUntil = %v, want the parent's %v", got.ValidUntil, now.Add(day))
		}
	})

	t.Run("an open-ended delegation inherits the parent's end", func(t *testing.T) {
		parent := activeFull
		parent.ValidUntil = ptr(now.Add(day))
		req := MandateGrant{Type: MandateAdministrative, Scope: MandateScopeOrganization, ValidFrom: now}
		got, err := clampToParent(req, parent, now)
		if err != nil {
			t.Fatalf("clampToParent: %v", err)
		}
		if got.ValidUntil == nil || !got.ValidUntil.Equal(now.Add(day)) {
			t.Errorf("validUntil = %v, want the parent's %v", got.ValidUntil, now.Add(day))
		}
	})

	t.Run("start is clamped up to the parent's start", func(t *testing.T) {
		parent := activeFull
		parent.ValidFrom = now.Add(-day)
		req := MandateGrant{
			Type: MandateAdministrative, Scope: MandateScopeOrganization,
			ValidFrom: now.Add(-10 * day),
		}
		got, err := clampToParent(req, parent, now)
		if err != nil {
			t.Fatalf("clampToParent: %v", err)
		}
		if !got.ValidFrom.Equal(parent.ValidFrom) {
			t.Errorf("validFrom = %v, want the parent's %v", got.ValidFrom, parent.ValidFrom)
		}
	})

	t.Run("a start after the parent's end leaves nothing to cut", func(t *testing.T) {
		parent := activeFull
		parent.ValidUntil = ptr(now.Add(day))
		req := MandateGrant{
			Type: MandateAdministrative, Scope: MandateScopeOrganization,
			ValidFrom: now.Add(2 * day),
		}
		if _, err := clampToParent(req, parent, now); !errors.Is(err, ErrOverDelegation) {
			t.Errorf("err = %v, want ErrOverDelegation", err)
		}
	})

	t.Run("a department parent cannot be widened to the org", func(t *testing.T) {
		parent := Mandate{
			Type: MandateFull, Scope: MandateScopeDepartment,
			ScopeDepartmentID: &dept, ValidFrom: now.Add(-day),
		}
		req := MandateGrant{Type: MandateAdministrative, Scope: MandateScopeOrganization, ValidFrom: now}
		if _, err := clampToParent(req, parent, now); !errors.Is(err, ErrOverDelegation) {
			t.Errorf("err = %v, want ErrOverDelegation", err)
		}
	})

	t.Run("a department parent cannot be moved to another department", func(t *testing.T) {
		parent := Mandate{
			Type: MandateFull, Scope: MandateScopeDepartment,
			ScopeDepartmentID: &dept, ValidFrom: now.Add(-day),
		}
		req := MandateGrant{
			Type: MandateAdministrative, Scope: MandateScopeDepartment,
			ScopeDepartmentID: &otherDept, ValidFrom: now,
		}
		if _, err := clampToParent(req, parent, now); !errors.Is(err, ErrOverDelegation) {
			t.Errorf("err = %v, want ErrOverDelegation", err)
		}
	})

	t.Run("an org-wide parent may be narrowed to a department", func(t *testing.T) {
		req := MandateGrant{
			Type: MandateAdministrative, Scope: MandateScopeDepartment,
			ScopeDepartmentID: &dept, ValidFrom: now,
		}
		if _, err := clampToParent(req, activeFull, now); err != nil {
			t.Errorf("clampToParent: %v", err)
		}
	})

	t.Run("nothing may be cut from an inactive parent", func(t *testing.T) {
		for name, parent := range map[string]Mandate{
			"revoked": {Type: MandateFull, Scope: MandateScopeOrganization, ValidFrom: now.Add(-day), RevokedAt: ptr(now.Add(-time.Hour))},
			"expired": {Type: MandateFull, Scope: MandateScopeOrganization, ValidFrom: now.Add(-2 * day), ValidUntil: ptr(now.Add(-day))},
			"pending": {Type: MandateFull, Scope: MandateScopeOrganization, ValidFrom: now.Add(day)},
		} {
			req := MandateGrant{Type: MandateAdministrative, Scope: MandateScopeOrganization, ValidFrom: now}
			if _, err := clampToParent(req, parent, now); !errors.Is(err, ErrMandateInactive) {
				t.Errorf("%s parent: err = %v, want ErrMandateInactive", name, err)
			}
		}
	})
}

func TestAuthorityWithdrawn(t *testing.T) {
	tests := []struct {
		name string
		a    Authority
		want bool
	}{
		{
			name: "an org that never granted a mandate is never withdrawn",
			a:    Authority{Granted: 0},
			want: false,
		},
		{
			name: "an active org-wide mandate is not withdrawn",
			a:    Authority{Granted: 1, Mandated: true},
			want: false,
		},
		{
			// Mandated counts only active org-wide grants, so this is equally the
			// caller whose every grant was revoked or expired and the one left with
			// nothing but a department-scoped mandate. Which of the two it is, is
			// ResolveAuthority's job and is covered in the store's integration test.
			name: "no active org-wide grant left is withdrawn",
			a:    Authority{Granted: 2, Mandated: false},
			want: true,
		},
		{
			name: "a platform admin is never withdrawn by an org's register",
			a:    Authority{Granted: 1, Mandated: false, PlatformAdmin: true},
			want: false,
		},
		{
			// Their authority is the register's, not a delegation, so a mandate they
			// happened to also be granted running out cannot take it away — the
			// register cannot withdraw the root it hangs from.
			name: "a legal representative is never withdrawn by the register they control",
			a:    Authority{Granted: 1, Mandated: false, LegalRepresentative: true},
			want: false,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := tt.a.Withdrawn(); got != tt.want {
				t.Errorf("Withdrawn() = %v, want %v", got, tt.want)
			}
		})
	}
}

func TestAuthorityMayGrantMandate(t *testing.T) {
	tests := []struct {
		name string
		a    Authority
		want bool
	}{
		{"legal representative may", Authority{LegalRepresentative: true}, true},
		{"full-mandate holder may", Authority{FullMandate: true, Mandated: true, Granted: 1}, true},
		{"an administrative mandate is not enough", Authority{Mandated: true, Granted: 1}, false},
		{"a platform admin is not the owner", Authority{PlatformAdmin: true}, false},
		{"nobody else may", Authority{}, false},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := tt.a.MayGrantMandate(); got != tt.want {
				t.Errorf("MayGrantMandate() = %v, want %v", got, tt.want)
			}
		})
	}
}

func TestValidateGrant(t *testing.T) {
	dept := uuid.New()
	now := time.Now()

	t.Run("an unknown tier is rejected", func(t *testing.T) {
		req := MandateGrant{Type: "owner", Scope: MandateScopeOrganization}
		if err := validateGrant(&req); !errors.Is(err, ErrMandateType) {
			t.Errorf("err = %v, want ErrMandateType", err)
		}
	})

	t.Run("a department scope needs a department", func(t *testing.T) {
		req := MandateGrant{Type: MandateFull, Scope: MandateScopeDepartment}
		if err := validateGrant(&req); !errors.Is(err, ErrMandateScope) {
			t.Errorf("err = %v, want ErrMandateScope", err)
		}
	})

	t.Run("an org scope may not carry a department", func(t *testing.T) {
		req := MandateGrant{Type: MandateFull, Scope: MandateScopeOrganization, ScopeDepartmentID: &dept}
		if err := validateGrant(&req); !errors.Is(err, ErrMandateScope) {
			t.Errorf("err = %v, want ErrMandateScope", err)
		}
	})

	t.Run("an unknown scope is rejected", func(t *testing.T) {
		req := MandateGrant{Type: MandateFull, Scope: "everything"}
		if err := validateGrant(&req); !errors.Is(err, ErrMandateScope) {
			t.Errorf("err = %v, want ErrMandateScope", err)
		}
	})

	t.Run("an empty window is rejected", func(t *testing.T) {
		req := MandateGrant{
			Type: MandateFull, Scope: MandateScopeOrganization,
			ValidFrom: now, ValidUntil: ptr(now),
		}
		if err := validateGrant(&req); !errors.Is(err, ErrMandateWindow) {
			t.Errorf("err = %v, want ErrMandateWindow", err)
		}
	})

	t.Run("a missing start defaults to now", func(t *testing.T) {
		req := MandateGrant{Type: MandateFull, Scope: MandateScopeOrganization}
		if err := validateGrant(&req); err != nil {
			t.Fatalf("validateGrant: %v", err)
		}
		if req.ValidFrom.IsZero() {
			t.Error("validFrom is still zero")
		}
	})
}

File: frontend/src/lib/mandate.test.ts
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { describe, expect, it } from "vitest";
import type { Mandate, MandateAuthority } from "../api/mandates";
import { en } from "../i18n/locales/en";
import {
  MANDATE_SCOPES,
  MANDATE_STATUSES,
  MANDATE_TYPES,
  MAX_REVOCATION_REASON_LENGTH,
  isoFromLocalInput,
  mandateCascade,
  mandateGrantAvailability,
  mandateIsRevocable,
  mandateLineage,
  mandateScopeLabel,
  mandateStatusTone,
  mandateTypeHint,
  mandateWindowIsEmpty,
} from "./mandate";

function mandate(overrides: Partial<Mandate> & { id: string }): Mandate {
  return {
    organizationId: "org-1",
    type: "full",
    status: "active",
    grantorUserId: "user-boss",
    grantorName: "Bo Boss",
    granteeUserId: `user-${overrides.id}`,
    granteeName: overrides.id,
    scope: "organization",
    scopeDepartmentId: null,
    scopeDepartmentName: null,
    parentMandateId: null,
    validFrom: "2026-01-01T00:00:00Z",
    validUntil: null,
    revokedAt: null,
    revokedByUserId: null,
    revocationReason: null,
    createdAt: "2026-01-01T00:00:00Z",
    ...overrides,
  };
}

const ids = (rows: { mandate: Mandate }[]): string[] =>
  rows.map((row) => row.mandate.id);
const depths = (rows: { depth: number }[]): number[] =>
  rows.map((row) => row.depth);

describe("mandateLineage", () => {
  it("keeps the order the backend served for a register with no delegations", () => {
    const rows = mandateLineage([
      mandate({ id: "c" }),
      mandate({ id: "a" }),
      mandate({ id: "b" }),
    ]);
    expect(ids(rows)).toEqual(["c", "a", "b"]);
    expect(depths(rows)).toEqual([0, 0, 0]);
  });

  it("puts a delegation under the mandate it was cut from", () => {
    // The register is served most recent first, so the child arrives first.
    const rows = mandateLineage([
      mandate({ id: "child", parentMandateId: "root" }),
      mandate({ id: "root" }),
    ]);
    expect(ids(rows)).toEqual(["root", "child"]);
    expect(depths(rows)).toEqual([0, 1]);
  });

  it("nests a whole chain, depth-first", () => {
    const rows = mandateLineage([
      mandate({ id: "grandchild", parentMandateId: "child" }),
      mandate({ id: "child", parentMandateId: "root" }),
      mandate({ id: "sibling", parentMandateId: "root" }),
      mandate({ id: "root" }),
      mandate({ id: "other-root" }),
    ]);
    expect(ids(rows)).toEqual([
      "root",
      "child",
      "grandchild",
      "sibling",
      "other-root",
    ]);
    expect(depths(rows)).toEqual([0, 1, 2, 1, 0]);
  });

  it("shows a mandate whose parent is not in the register as a root", () => {
    // The register is the audit surface: a row that vanished from it would hide
    // a grant, so an unresolvable parent costs the indent, not the row.
    const rows = mandateLineage([
      mandate({ id: "orphan", parentMandateId: "gone" }),
    ]);
    expect(ids(rows)).toEqual(["orphan"]);
    expect(depths(rows)).toEqual([0]);
  });

  it("still lists every mandate when the parent links form a cycle", () => {
    // Nothing the backend serves can do this (a parent always predates its
    // child), but neither row may drop out of the register if it ever does.
    const rows = mandateLineage([
      mandate({ id: "a", parentMandateId: "b" }),
      mandate({ id: "b", parentMandateId: "a" }),
    ]);
    expect(ids(rows).toSorted()).toEqual(["a", "b"]);
  });
});

describe("mandateCascade", () => {
  const register = [
    mandate({ id: "root" }),
    mandate({ id: "child", parentMandateId: "root" }),
    mandate({
      id: "pending-child",
      parentMandateId: "root",
      status: "pending",
    }),
    mandate({ id: "grandchild", parentMandateId: "child" }),
    mandate({ id: "elsewhere" }),
  ];

  it("reaches every descendant of the mandate being revoked", () => {
    expect(mandateCascade(register, "root").map((m) => m.id)).toEqual([
      "child",
      "pending-child",
      "grandchild",
    ]);
  });

  it("does not count the mandate itself, nor an unrelated one", () => {
    const reached = mandateCascade(register, "root").map((m) => m.id);
    expect(reached).not.toContain("root");
    expect(reached).not.toContain("elsewhere");
  });

  it("is empty for a mandate nothing was delegated from", () => {
    expect(mandateCascade(register, "grandchild")).toEqual([]);
  });

  it("leaves out a descendant that already ended, matching the backend", () => {
    // The backend walks the whole subtree but only touches rows that are neither
    // revoked nor expired, so counting an ended one would overstate the warning.
    const ended = [
      mandate({ id: "root" }),
      mandate({ id: "revoked", parentMandateId: "root", status: "revoked" }),
      mandate({ id: "expired", parentMandateId: "root", status: "expired" }),
    ];
    expect(mandateCascade(ended, "root")).toEqual([]);
  });

  it("keeps descending past an ended descendant to the live ones under it", () => {
    // The recursive CTE walks by parent id alone, so a live grandchild under a
    // revoked child is still revoked with it.
    const register = [
      mandate({ id: "root" }),
      mandate({ id: "revoked", parentMandateId: "root", status: "revoked" }),
      mandate({ id: "under-it", parentMandateId: "revoked" }),
    ];
    expect(mandateCascade(register, "root").map((m) => m.id)).toEqual([
      "under-it",
    ]);
  });
});

describe("mandateGrantAvailability", () => {
  const authority = (
    overrides: Partial<MandateAuthority> = {},
  ): MandateAuthority => ({
    mayGrant: false,
    legalRepresentative: false,
    fullMandate: false,
    jointAuthority: false,
    ...overrides,
  });

  it("offers nothing to a caller the backend would refuse", () => {
    expect(mandateGrantAvailability(authority())).toBe("noAuthority");
  });

  it("offers the flows to a legal representative and to a full-mandate holder", () => {
    expect(
      mandateGrantAvailability(
        authority({ mayGrant: true, legalRepresentative: true }),
      ),
    ).toBe("available");
    expect(
      mandateGrantAvailability(
        authority({ mayGrant: true, fullMandate: true }),
      ),
    ).toBe("available");
  });

  it("withholds them from a jointly registered director", () => {
    // The backend would accept the grant, because it does not read the authority
    // column yet; offering the flow would record a grant one director cannot make.
    expect(
      mandateGrantAvailability(
        authority({
          mayGrant: true,
          legalRepresentative: true,
          jointAuthority: true,
        }),
      ),
    ).toBe("jointAuthority");
  });

  it("lifts the hold when the joint director also holds a full mandate", () => {
    // That mandate is a basis of its own, granted through the register rather
    // than read off it, so acting on it is not acting alone as a director.
    expect(
      mandateGrantAvailability(
        authority({
          mayGrant: true,
          legalRepresentative: true,
          jointAuthority: true,
          fullMandate: true,
        }),
      ),
    ).toBe("available");
  });
});

describe("mandateIsRevocable", () => {
  it("covers the statuses that still have a life ahead of them", () => {
    expect(mandateIsRevocable(mandate({ id: "a", status: "active" }))).toBe(
      true,
    );
    expect(mandateIsRevocable(mandate({ id: "a", status: "pending" }))).toBe(
      true,
    );
  });

  it("is false for a mandate that ended, which the backend answers 409 for", () => {
    expect(mandateIsRevocable(mandate({ id: "a", status: "revoked" }))).toBe(
      false,
    );
    expect(mandateIsRevocable(mandate({ id: "a", status: "expired" }))).toBe(
      false,
    );
  });
});

describe("mandateStatusTone", () => {
  it("badges each status", () => {
    expect(mandateStatusTone("active")).toBe("green");
    expect(mandateStatusTone("pending")).toBe("amber");
    expect(mandateStatusTone("revoked")).toBe("red");
    expect(mandateStatusTone("expired")).toBe("default");
  });

  it("falls back to the neutral tone for a status it does not know", () => {
    expect(mandateStatusTone("something-new")).toBe("default");
  });
});

describe("mandateScopeLabel", () => {
  const t = ((key: string) => key) as unknown as Parameters<
    typeof mandateScopeLabel
  >[1];

  it("names the department a department-scoped mandate is confined to", () => {
    expect(
      mandateScopeLabel(
        mandate({
          id: "a",
          scope: "department",
          scopeDepartmentId: "dept-1",
          scopeDepartmentName: "Finance",
        }),
        t,
      ),
    ).toBe("Finance");
  });

  it("falls back to the generic label when the department has no name", () => {
    expect(
      mandateScopeLabel(mandate({ id: "a", scope: "department" }), t),
    ).toBe("mandates.scopes.department");
  });

  it("labels an org-wide mandate", () => {
    expect(mandateScopeLabel(mandate({ id: "a" }), t)).toBe(
      "mandates.scopes.organization",
    );
  });
});

describe("mandateWindowIsEmpty", () => {
  const now = "2026-06-01T12:00:00.000Z";

  it("accepts an open-ended window", () => {
    expect(mandateWindowIsEmpty(undefined, undefined, now)).toBe(false);
    expect(
      mandateWindowIsEmpty("2026-07-01T00:00:00.000Z", undefined, now),
    ).toBe(false);
  });

  it("rejects an end at or before the start named", () => {
    expect(
      mandateWindowIsEmpty(
        "2026-07-01T00:00:00.000Z",
        "2026-06-15T00:00:00.000Z",
        now,
      ),
    ).toBe(true);
    expect(
      mandateWindowIsEmpty(
        "2026-07-01T00:00:00.000Z",
        "2026-07-01T00:00:00.000Z",
        now,
      ),
    ).toBe(true);
  });

  it("rejects an end in the past when no start is named", () => {
    // validateGrant fills valid_from with now, so this is the empty window the
    // backend answers 400 for; leaving it to the round trip shows its
    // untranslated prose.
    expect(
      mandateWindowIsEmpty(undefined, "2026-05-01T00:00:00.000Z", now),
    ).toBe(true);
    expect(mandateWindowIsEmpty(undefined, now, now)).toBe(true);
  });

  it("accepts an end in the future when no start is named", () => {
    expect(
      mandateWindowIsEmpty(undefined, "2026-07-01T00:00:00.000Z", now),
    ).toBe(false);
  });
});

describe("mandateTypeHint", () => {
  const t = ((key: string) => key) as unknown as Parameters<
    typeof mandateTypeHint
  >[1];

  it("explains each tier it knows", () => {
    expect(mandateTypeHint("full", t)).toBe("mandates.typeHints.full");
    expect(mandateTypeHint("administrative", t)).toBe(
      "mandates.typeHints.administrative",
    );
  });

  it("has no hint for a tier it does not know", () => {
    expect(mandateTypeHint("something-new", t)).toBeUndefined();
  });
});

describe("isoFromLocalInput", () => {
  it("has no value for an empty field, so the backend default applies", () => {
    expect(isoFromLocalInput("")).toBeUndefined();
    expect(isoFromLocalInput("   ")).toBeUndefined();
  });

  it("turns a datetime-local value into an instant", () => {
    // The field carries no zone, so it is read as local time; comparing against
    // Date keeps the test independent of the runner's timezone.
    expect(isoFromLocalInput("2026-06-01T12:30")).toBe(
      new Date("2026-06-01T12:30").toISOString(),
    );
  });

  it("has no value for something unparseable", () => {
    expect(isoFromLocalInput("not-a-date")).toBeUndefined();
  });
});

// The backend is the source of truth for the mandate tiers, scopes and statuses
// (backend/internal/organization/mandate.go). Every one of them reaches the
// register screen as a label, and the grant form offers the tiers from
// MANDATE_TYPES, so a tier the backend gains and this file omits would be offered
// to nobody and would render as a bare "full" in the table. This test parses the
// Go constants and asserts membership in both directions, plus that each value is
// named in en.ts. nl.ts is typed against en.ts, so the typecheck already fails on
// a missing Dutch twin.

const mandateGoPath = fileURLToPath(
  new URL("../../../backend/internal/organization/mandate.go", import.meta.url),
);
const mandateSource = readFileSync(mandateGoPath, "utf8");

const handlerGoPath = fileURLToPath(
  new URL(
    "../../../backend/internal/organization/mandates.go",
    import.meta.url,
  ),
);
const handlerSource = readFileSync(handlerGoPath, "utf8");

// The type is optional in each pattern on purpose: inside a `const` block a later
// member may be written without repeating it and is still an untyped string
// constant the API can serve.
function goStringConstants(prefix: string): string[] {
  const pattern = new RegExp(
    `^\\s*${prefix}\\w+(?:\\s+\\w+)?\\s*=\\s*"([^"]+)"`,
    "gm",
  );
  return [...mandateSource.matchAll(pattern)].map((match) => match[1]);
}

const backendStatuses = goStringConstants("MandateStatus");
const backendScopes = goStringConstants("MandateScope");
const backendTypes = goStringConstants("Mandate(?!Scope|Status)");

describe("mandate constants backend/frontend parity", () => {
  it("extracts the constants from mandate.go", () => {
    expect(backendTypes).toContain("full");
    expect(backendScopes).toContain("organization");
    expect(backendStatuses).toContain("active");
  });

  it.each(backendTypes)("accepts and names the tier %s", (type) => {
    expect(MANDATE_TYPES).toContain(type);
    expect(
      en.mandates.types[type as keyof typeof en.mandates.types],
    ).toBeTruthy();
    expect(
      en.mandates.typeHints[type as keyof typeof en.mandates.typeHints],
    ).toBeTruthy();
    // Naming the hint in en.ts is not enough: the grant form reads it through
    // TYPE_HINT_KEYS, so a tier missing from that map would go unexplained.
    expect(mandateTypeHint(type, ((key: string) => key) as never)).toBe(
      `mandates.typeHints.${type}`,
    );
  });

  it.each([...MANDATE_TYPES])("is served the tier %s", (type) => {
    expect(backendTypes).toContain(type);
  });

  it.each(backendScopes)("accepts and names the scope %s", (scope) => {
    expect(MANDATE_SCOPES).toContain(scope);
    expect(
      en.mandates.scopes[scope as keyof typeof en.mandates.scopes],
    ).toBeTruthy();
  });

  it.each([...MANDATE_SCOPES])("is served the scope %s", (scope) => {
    expect(backendScopes).toContain(scope);
  });

  it.each(backendStatuses)("accepts and names the status %s", (status) => {
    expect(MANDATE_STATUSES).toContain(status);
    expect(
      en.mandates.statuses[status as keyof typeof en.mandates.statuses],
    ).toBeTruthy();
  });

  it.each([...MANDATE_STATUSES])("is served the status %s", (status) => {
    expect(backendStatuses).toContain(status);
  });

  it("bounds the revocation reason at the length the backend accepts", () => {
    const found = /maxRevocationReasonLength\s*=\s*(\d+)/.exec(
      handlerSource,
    )?.[1];
    expect(found).toBe(String(MAX_REVOCATION_REASON_LENGTH));
  });
});
